### PR TITLE
convert EmptyState to ts

### PR DESCRIPTION
**Pull request by danethurber**
- [x] convert EmptyState to ts
- [x] convert core to ts
- [x] convert util to ts
- [x] move common utils/types to central place

### CODE REVIEW

File: packages/emptystate/src/react/index.tsx [hunk 1]
```diff
@@ -0,0 +1,202 @@
+import { StyleAttribute, compose, css } from 'glamor'
```
Comment (danethurber) [inline packages/emptystate/src/react/index.tsx:1]: ugh....why are my git mv commands not working 🤮

Comment (jaketrent) [reply packages/emptystate/src/react/index.tsx:1]: dunno.  

fwiw, I usually don't `git mv`.  I just `mv` and then `git add`, and then in the commit, it's seen as a move.

File: jest/setup/index.js [hunk 2]
```diff
@@ -1,7 +1,10 @@
...
 import 'jest-prop-type-error'
 
-import '@testing-library/jest-dom/extend-expect.js'
+import '@testing-library/jest-dom/extend-expect'
 
 import './match-media.js'
 import './react-dom.js'
+
+import addons, { mockChannel } from '@storybook/addons'
+addons.setChannel(mockChannel())
```
Comment (danethurber) [inline jest/setup/index.js:10]: needed if you import a story to be used in a jest test

File: packages/emptystate/src/react/index.tsx [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  const { children: custom, name, ...rest } = props
+
+  return (
+    <Context.Consumer>
+      {ctx => {
+        const Comp = illustrations[name] || IllustrationNotFound
+        // const isSmall: boolean = ctx.size === sizes.small && !!Comp.small
+
+        // if (isSmall) Comp = Comp.small
+        // if (custom) Comp = () => custom
```
Comment (danethurber) [inline packages/emptystate/src/react/index.tsx:175]: oops. looks like i still need to fix this

File: packages/emptystate/src/vars/index.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+type KeyMirror<Keys extends string[]> = { readonly [K in Keys[number]]: K }
+
+function keyMirror<Keys extends string[]>(...inputs: Keys) {
+  const mirrored = inputs.reduce(
+    (acc, key) => ({ ...acc, [key]: key }),
+    {} as KeyMirror<Keys>
+  )
+
+  return Object.freeze(mirrored)
+}
```
Comment (danethurber) [inline packages/emptystate/src/vars/index.ts:30]: looks like i missed this. will update

File: packages/util/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+export { capitalize } from './string'
+export { stylesFor } from './styles'
+export { useCombinedRefs } from './use-combined-refs'
+export { usePrevious } from './use-previous'
+export { useResizeObserver } from './use-resize-observer'
+export { useToggle } from './use-toggle'
+export {
+  useCloseOnDocumentEvents,
+  onGlobalEventsClose
+} from './use-close-on-document-events'
```
Comment (danethurber) [inline packages/util/src/index.ts:24]: i converted some of the util functions to typescript but not all. we'll need to followup later if we want everything in ts

File: packages/util/src/menu-keys/__specs__/index.spec.tsx [hunk 6]
```diff
@@ -69,8 +11,9 @@ describe('useMenuKeyEvents()', () => {
       }
     })
   })
+
   it('ArrowUp', () => {
-    const { getByTestId } = render(<TestComponent />)
+    const { getByTestId } = render(basic())
```
Comment (danethurber) [inline packages/util/src/menu-keys/__specs__/index.spec.tsx:16]: example of how to use a story in a jest test

**Review by jaketrent — APPROVED** [review]

File: .eslintrc.js [hunk 7]
```diff
@@ -63,33 +63,35 @@ module.exports = {
     '@typescript-eslint/restrict-plus-operands': WARNING,
     '@typescript-eslint/restrict-template-expressions': WARNING,
     '@typescript-eslint/unbound-method': WARNING,
+    'no-unused-expressions': IGNORE,
+    'no-use-before-define': IGNORE,
     'prettier/prettier': ERROR,
     'react-hooks/exhaustive-deps': WARNING,
     'react-hooks/rules-of-hooks': ERROR,
-    'react/display-name': WARNING,
+    'react/display-name': IGNORE,
```
Comment (jaketrent) [inline .eslintrc.js:71]: too many?

**Review by jaketrent — COMMENTED** [review]
yarn build gives me:

```
src/css/index.ts:73:46 - error TS2339: Property 'small' does not exist on type '{ [x: number]: string; }'.

73   ['.psds-emptystate__heading' + sizeClasses.small]: {
                                                ~~~~~

src/css/index.ts:76:46 - error TS2339: Property 'large' does not exist on type '{ [x: number]: string; }'.

76   ['.psds-emptystate__heading' + sizeClasses.large]: {
                                                ~~~~~

src/css/index.ts:87:51 - error TS2339: Property 'small' does not exist on type '{ [x: number]: string; }'.

87   ['.psds-emptystate__illustration' + sizeClasses.small]: {
                                                     ~~~~~

src/css/index.ts:91:51 - error TS2339: Property 'large' does not exist on type '{ [x: number]: string; }'.

91   ['.psds-emptystate__illustration' + sizeClasses.large]: {
                                                     ~~~~~

src/react/index.tsx:187:12 - error TS2339: Property 'Actions' does not exist on type 'EmptyStateComponent'.

187 EmptyState.Actions = Actions
               ~~~~~~~

src/react/index.tsx:188:12 - error TS2339: Property 'Caption' does not exist on type 'EmptyStateComponent'.

188 EmptyState.Caption = Caption
               ~~~~~~~

src/react/index.tsx:189:12 - error TS2339: Property 'Heading' does not exist on type 'EmptyStateComponent'.

189 EmptyState.Heading = Heading
               ~~~~~~~

src/react/index.tsx:190:12 - error TS2339: Property 'Illustration' does not exist on type 'EmptyStateComponent'.

190 EmptyState.Illustration = Illustration
               ~~~~~~~~~~~~

src/react/index.tsx:192:12 - error TS2339: Property 'sizes' does not exist on type 'EmptyStateComponent'.

192 EmptyState.sizes = sizes
               ~~~~~


Found 9 errors.
```

Should I be concerned?  Or just add an eslint ignore?

**Review by EdwardIrby — APPROVED** [review]

**Comment by danethurber** [comment]
@jaketrent errors fixed and linting updated

**Pull request merged**